File: Components/WinForms/CS/GridMainDemo/Modules/Tasks.cs
using System;
using System.Collections;
using System.ComponentModel;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Columns;

namespace DevExpress.XtraGrid.Demos {
    public abstract class BindingCollection : CollectionBase, IBindingList {
        public virtual void OnListChanged(object item) { }
        public int IndexOf(object item) {
            return List.IndexOf(item);
        }
        public object AddNew() { return null; }
        public bool AllowEdit { get { return true; } }
        public bool AllowNew { get { return false; } }
        public bool AllowRemove { get { return true; } }

        private ListChangedEventHandler listChangedHandler;
        public event ListChangedEventHandler ListChanged {
            add { listChangedHandler += value; }
            remove { listChangedHandler -= value; }
        }
        internal void OnListChanged(ListChangedEventArgs args) {
            if(listChangedHandler != null) {
                listChangedHandler(this, args);
            }
        }
        protected override void OnRemoveComplete(int index, object value) {
            OnListChanged(new ListChangedEventArgs(ListChangedType.ItemDeleted, index));
        }
        protected override void OnInsertComplete(int index, object value) {
            OnListChanged(new ListChangedEventArgs(ListChangedType.ItemAdded, index));
        }

        public void AddIndex(PropertyDescriptor pd) { throw new NotSupportedException(); }
        public void ApplySort(PropertyDescriptor pd, ListSortDirection dir) { throw new NotSupportedException(); }
        public int Find(PropertyDescriptor property, object key) { throw new NotSupportedException(); }
        public bool IsSorted { get { return false; } }
        public void RemoveIndex(PropertyDescriptor pd) { throw new NotSupportedException(); }
        public void RemoveSort() { throw new NotSupportedException(); }
        public ListSortDirection SortDirection { get { throw new NotSupportedException(); } }
        public PropertyDescriptor SortProperty { get { throw new NotSupportedException(); } }
        public bool SupportsChangeNotification { get { return true; } }
        public bool SupportsSearching { get { return false; } }
        public bool SupportsSorting { get { return false; } }
    }

    public enum Priority { Low, Medium, High }

    public class Task {
        int fID;
        string fName;
        DateTime fDate;
        int fPercentComplete;
        bool fComplete;
        string fNote;
        Priority fPriority;

        BindingCollection fRelationCollection;
        public Task(BindingCollection relationCollection, int id, string name, DateTime date) {
            this.fRelationCollection = relationCollection;
            this.fID = id;
            this.fName = name;
            this.fDate = date;
            this.fPercentComplete = 0;
            this.fComplete = false;
            this.fNote = "";
            this.fPriority = Priority.Low;
        }

        public int ID {
            get { return fID; }
        }

        public string TaskName {
            get { return fName; }
            set {
                fName = value;
                OnListChanged();
            }
        }

        public Priority Priority {
            get { return fPriority; }
            set {
                fPriority = value;
                OnListChanged();
            }
        }

        public DateTime DueDate {
            get { return fDate; }
            set {
                fDate = value;
                OnListChanged();
            }
        }

        public bool Complete {
            get { return fComplete; }
            set {
                fComplete = value;
                if(fComplete) fPercentComplete = 100;
                if(!fComplete && fPercentComplete == 100) fPercentComplete = 0;
                OnListChanged();
            }
        }

        public int PercentComplete {
            get { return fPercentComplete; }
            set {
                fPercentComplete = value;
                if(fPercentComplete < 0) fPercentComplete = 0;
                if(fPercentComplete > 100) fPercentComplete = 100;
                fComplete = fPercentComplete == 100;
                OnListChanged();
            }
        }

        public string Note {
            get { return fNote; }
            set {
                fNote = value;
                OnListChanged();
            }
        }

        public string CategoryName {
            get { return GetCategoryByTask((TasksWithCategories)fRelationCollection, this); }
        }

        static string GetCategoryByTask(TasksWithCategories collection, Task task) {
            string ret = "";
            for(int i = 0; i < collection.fCategories.Count; i++) {
                if(collection.HasCategory(task, collection.fCategories[i]))
                    ret += string.Format("{0}{1}", (ret == "" ? "" : ", "), collection.fCategories[i].CategoryName);
            }
            if(ret == "") ret = Properties.Resources.NoneString;
            return ret;
        }

        private void OnListChanged() {
            fRelationCollection.OnListChanged(this);
        }
    }
    public class Category {
        int fID;
        string fName;
        public Category(int id, string name) {
            this.fID = id;
            this.fName = name;
        }

        public int ID {
            get { return fID; }
        }

        public string CategoryName {
            get { return fName; }
            set { fName = value; }
        }
    }
    public class Relation {
        internal Task fTask;
        internal Category fCategory;
        public Relation(Task task, Category category) {
            this.fTask = task;
            this.fCategory = category;
        }
        public string TaskName {
            get { return fTask.TaskName; }
            set { fTask.TaskName = value; }
        }
        public DateTime DueDate {
            get { return fTask.DueDate; }
            set { fTask.DueDate = value; }
        }
        public string CategoryName {
            get { return fCategory.CategoryName; }
        }
        public bool Complete {
            get { return fTask.Complete; }
            set { fTask.Complete = value; }
        }

        public int PercentComplete {
            get { return fTask.PercentComplete; }
            set { fTask.PercentComplete = value; }
        }

        public string Note {
            get { return fTask.Note; }
            set { fTask.Note = value; }
        }
        public Priority Priority {
            get { return fTask.Priority; }
            set { fTask.Priority = value; }
        }
    }
    public class Tasks : BindingCollection {
        public static int MaxTasks = 7;
        public static Tasks GetTasksWithPriority(TasksWithCategories collection) {
            Tasks ret = new Tasks();
            Random rnd = new Random();
            string[] names = new string[] { 
                "Create the gauges demo design",
                "Review the new Layout Form",
                "Check XAF Wizard's new pages",
                "Create icons for the Layout Control's context menu",
                "Create Bizcard Design for the blog post",
                "Create Booth design for TechEd 2015",
                "Add grayscale icons to the Image Gallery"};
            for(int i = 0; i < MaxTasks; i++) {
                ret.List.Add(new Task(collection, i + 1, Properties.Resources.Task + (i + 1).ToString(), DateTime.Today.AddDays(rnd.Next(-3, 5))));
                ret[i].TaskName = names[i];
                ret[i].PercentComplete = rnd.Next(1, 99);
                if(i == 1) { ret[i].Priority = Priority.Medium; }
                if(i == 2) { ret[i].Priority = Priority.Medium; }
                if(i == 6) { ret[i].Priority = Priority.High; }
            }
            return ret;
        }
        public static Tasks GetTasks(TasksWithCategories collection) {
            Tasks ret = new Tasks();
            Random rnd = new Random();
            for(int i = 0; i < MaxTasks; i++) {
                ret.List.Add(new Task(collection, i + 1, Properties.Resources.Task + (i + 1).ToString(), DateTime.Today.AddDays(rnd.Next(5))));
                if(i == 2) { ret[i].PercentComplete = 50; }
                if(i == 6) { ret[i].PercentComplete = 100; }
            }
            return ret;
        }
        public Task this[int index] {
            get { return (Task)(List[index]); }
            set { List[index] = value; }
        }
    }
    public class Categories : BindingCollection {
        public static int MaxCategories = 6;
        public static Categories GetCategories(TasksWithCategories collection) {
            Categories ret = new Categories();
            string[] names = new string[] {Properties.Resources.Business, Properties.Resources.Competitor, Properties.Resources.Favorites,
                Properties.Resources.Gifts, Properties.Resources.Goals, Properties.Resources.Holiday, 
                Properties.Resources.Ideas, Properties.Resources.International, Properties.Resources.Personal};
            for(int i = 0; i < names.Length; i++)
                ret.List.Add(new Category(i + 1, names[i]));
            return ret;
        }
        public Category this[int index] {
            get { return (Category)(List[index]); }
            set { List[index] = value; }
        }
    }
    public class TasksWithCategories : BindingCollection {
        internal Tasks fTasks;
        internal Categories fCategories;
        public TasksWithCategories() {
            fTasks = Tasks.GetTasks(this);
            fCategories = Categories.GetCategories(this);
        }
        public TasksWithCategories(bool usePriorities) {
            if(usePriorities)
                fTasks = Tasks.GetTasksWithPriority(this);
            else fTasks = Tasks.GetTasks(this);
            fCategories = Categories.GetCategories(this);
        }
        public static TasksWithCategories GetTasksWithCategories() {
            TasksWithCategories ret = new TasksWithCategories();
            Random rnd = new Random();
            for(int i = 0; i < Tasks.MaxTasks; i++)
                for(int j = 0; j < 1 + rnd.Next(Categories.MaxCategories); j++) {
                    Category cat = ret.fCategories[rnd.Next(ret.fCategories.Count)];
                    if(!ret.HasCategory(ret.fTasks[i], cat))
                        ret.List.Add(new Relation(ret.fTasks[i], cat));
                }
            return ret;
        }
        public static TasksWithCategories GetTasksWithPriorities() {
            TasksWithCategories ret = new TasksWithCategories(true);
            Random rnd = new Random();
            for(int i = 0; i < Tasks.MaxTasks; i++) {
                Category cat = ret.fCategories[rnd.Next(ret.fCategories.Count)];
                if(!ret.HasCategory(ret.fTasks[i], cat))
                    ret.List.Add(new Relation(ret.fTasks[i], cat));
            }
            return ret;
        }
        public Relation this[int index] {
            get { return (Relation)(List[index]); }
            set { List[index] = value; }
        }
        public override void OnListChanged(object item) {
            if(item == null) return;
            for(int i = 0; i < this.Count; i++)
                if(item.Equals(this[i].fTask))
                    this.OnListChanged(new ListChangedEventArgs(ListChangedType.ItemChanged, i));
        }
        public bool HasCategory(Task task, Category category) {
            for(int i = 0; i < this.Count; i++)
                if(this[i].fCategory == category && this[i].fTask == task)
                    return true;
            return false;
        }
    }
    //<gridControl1>
    public class TasksWithCategoriesDatasource {
        GridControl fGrid;
        TasksWithCategories fTasks;
        public TasksWithCategoriesDatasource(GridControl grid) {
            this.fGrid = grid;
            this.fTasks = TasksWithCategories.GetTasksWithPriorities();
            GridView view = fGrid.MainView as GridView;
            SetDataSource();
        }
        public void SetDataSource() {
            fGrid.DataSource = fTasks.fTasks;
        }
    }
    public class GroupingControllerTasksWithCategories {
        GridControl fGrid;
        TasksWithCategories fTasks;
        public event EventHandler AfterGrouping;
        public GroupingControllerTasksWithCategories(GridControl grid) {
            this.fGrid = grid;
            this.fTasks = TasksWithCategories.GetTasksWithCategories();
            GridView view = fGrid.MainView as GridView;
            if(view != null)
                view.EndGrouping += new EventHandler(Grid_Grouping);
            SetDataSource();
        }
        void Grid_Grouping(object sender, EventArgs e) {
            SetDataSource();
            if(AfterGrouping != null) AfterGrouping(this, EventArgs.Empty);
        }

        public GridColumn CategoryColumn {
            get {
                GridView view = fGrid.MainView as GridView;
                if(view == null) return null;
                foreach(GridColumn column in view.Columns)
                    if(column.FieldName == "CategoryName") return column;
                return null;
            }
        }

        public bool IsCategoryGrouping {
            get {
                if(CategoryColumn == null) return false;
                return CategoryColumn.GroupIndex > -1;
            }
        }

        public void SetDataSource() {
            if(IsCategoryGrouping)
                fGrid.DataSource = fTasks;
            else fGrid.DataSource = fTasks.fTasks;
        }
    }
    //</gridControl1>

}

File: Components/WPF/VB/PivotGridDemo.Wpf/Controls/Tasks.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Data
Imports System.Windows.Documents
Imports System.Windows.Input
Imports System.Windows.Media
Imports System.Windows.Media.Imaging
Imports System.Windows.Shapes
Imports System.Xml
Imports DevExpress.Data.Filtering
Imports System.Data.SqlClient
Imports System.Data
Imports System.Windows.Threading
Imports System.Threading
Imports System.ComponentModel
Imports DevExpress.XtraEditors.DXErrorProvider
Imports DevExpress.Data

Namespace PivotGridDemo.PivotGrid
    Public MustInherit Class BindingCollection
        Inherits CollectionBase
        Implements IBindingList

        Public Overridable Sub OnListChanged(ByVal item As Object)
        End Sub
        Public Function AddNew() As Object Implements IBindingList.AddNew
            Return Nothing
        End Function
        Public ReadOnly Property AllowEdit() As Boolean Implements IBindingList.AllowEdit
            Get
                Return True
            End Get
        End Property
        Public ReadOnly Property AllowNew() As Boolean Implements IBindingList.AllowNew
            Get
                Return False
            End Get
        End Property
        Public ReadOnly Property AllowRemove() As Boolean Implements IBindingList.AllowRemove
            Get
                Return True
            End Get
        End Property

        Private listChangedHandler As ListChangedEventHandler
        Public Custom Event ListChanged As ListChangedEventHandler Implements IBindingList.ListChanged
            AddHandler(ByVal value As ListChangedEventHandler)
                listChangedHandler = DirectCast(System.Delegate.Combine(listChangedHandler, value), ListChangedEventHandler)
            End AddHandler
            RemoveHandler(ByVal value As ListChangedEventHandler)
                listChangedHandler = DirectCast(System.Delegate.Remove(listChangedHandler, value), ListChangedEventHandler)
            End RemoveHandler
            RaiseEvent(ByVal sender As System.Object, ByVal e As System.ComponentModel.ListChangedEventArgs)
                If listChangedHandler IsNot Nothing Then
                    For Each d As ListChangedEventHandler In listChangedHandler.GetInvocationList()
                        d.Invoke(sender, e)
                    Next d
                End If
            End RaiseEvent
        End Event
        Friend Sub OnListChanged(ByVal args As ListChangedEventArgs)
            If listChangedHandler IsNot Nothing Then
                listChangedHandler(Me, args)
            End If
        End Sub
        Protected Overrides Sub OnRemoveComplete(ByVal index As Integer, ByVal value As Object)
            OnListChanged(New ListChangedEventArgs(ListChangedType.ItemDeleted, index))
        End Sub
        Protected Overrides Sub OnInsertComplete(ByVal index As Integer, ByVal value As Object)
            OnListChanged(New ListChangedEventArgs(ListChangedType.ItemAdded, index))
        End Sub

        Public Sub AddIndex(ByVal pd As PropertyDescriptor) Implements IBindingList.AddIndex
            Throw New NotSupportedException()
        End Sub
        Public Sub ApplySort(ByVal pd As PropertyDescriptor, ByVal dir As ListSortDirection) Implements IBindingList.ApplySort
            Throw New NotSupportedException()
        End Sub
        Public Function Find(ByVal [property] As PropertyDescriptor, ByVal key As Object) As Integer Implements IBindingList.Find
            Throw New NotSupportedException()
        End Function
        Public ReadOnly Property IsSorted() As Boolean Implements IBindingList.IsSorted
            Get
                Return False
            End Get
        End Property
        Public Sub RemoveIndex(ByVal pd As PropertyDescriptor) Implements IBindingList.RemoveIndex
            Throw New NotSupportedException()
        End Sub
        Public Sub RemoveSort() Implements IBindingList.RemoveSort
            Throw New NotSupportedException()
        End Sub
        Public ReadOnly Property SortDirection() As ListSortDirection Implements IBindingList.SortDirection
            Get
                Throw New NotSupportedException()
            End Get
        End Property
        Public ReadOnly Property SortProperty() As PropertyDescriptor Implements IBindingList.SortProperty
            Get
                Throw New NotSupportedException()
            End Get
        End Property
        Public ReadOnly Property SupportsChangeNotification() As Boolean Implements IBindingList.SupportsChangeNotification
            Get
                Return True
            End Get
        End Property
        Public ReadOnly Property SupportsSearching() As Boolean Implements IBindingList.SupportsSearching
            Get
                Return False
            End Get
        End Property
        Public ReadOnly Property SupportsSorting() As Boolean Implements IBindingList.SupportsSorting
            Get
                Return False
            End Get
        End Property
    End Class
    Public Class Task
        Private fID As Integer
        Private fName As String
        Private fDate As Date
        Private fPercentComplete As Integer
        Private fComplete As Boolean
        Private fNote As String

        Private fRelationCollection As BindingCollection
        Public Sub New(ByVal relationCollection As BindingCollection, ByVal id As Integer, ByVal name As String, ByVal [date] As Date)
            Me.fRelationCollection = relationCollection
            Me.fID = id
            Me.fName = name
            Me.fDate = [date]
            Me.fPercentComplete = 0
            Me.fComplete = False
            Me.fNote = ""
        End Sub

        Public ReadOnly Property ID() As Integer
            Get
                Return fID
            End Get
        End Property

        Public Property TaskName() As String
            Get
                Return fName
            End Get
            Set(ByVal value As String)
                fName = value
                OnListChanged()
            End Set
        End Property

        Public Property DueDate() As Date
            Get
                Return fDate
            End Get
            Set(ByVal value As Date)
                fDate = value
                OnListChanged()
            End Set
        End Property

        Public Property Complete() As Boolean
            Get
                Return fComplete
            End Get
            Set(ByVal value As Boolean)
                fComplete = value
                If fComplete Then
                    fPercentComplete = 100
                End If
                If (Not fComplete) AndAlso fPercentComplete = 100 Then
                    fPercentComplete = 0
                End If
                OnListChanged()
            End Set
        End Property

        Public Property PercentComplete() As Integer
            Get
                Return fPercentComplete
            End Get
            Set(ByVal value As Integer)
                fPercentComplete = value
                If fPercentComplete < 0 Then
                    fPercentComplete = 0
                End If
                If fPercentComplete > 100 Then
                    fPercentComplete = 100
                End If
                fComplete = fPercentComplete = 100
                OnListChanged()
            End Set
        End Property

        Public Property Note() As String
            Get
                Return fNote
            End Get
            Set(ByVal value As String)
                fNote = value
                OnListChanged()
            End Set
        End Property

        Public ReadOnly Property CategoryName() As String
            Get
                Return GetCategoryByTask(DirectCast(fRelationCollection, TasksWithCategories), Me)
            End Get
        End Property

        Private Shared Function GetCategoryByTask(ByVal collection As TasksWithCategories, ByVal task As Task) As String
            Dim ret As String = ""
            For i As Integer = 0 To collection.fCategories.Count - 1
                If collection.HasCategory(task, collection.fCategories(i)) Then
                    ret &= String.Format("{0}{1}", (If(ret = "", "", ", ")), collection.fCategories(i).CategoryName)
                End If
            Next i
            If ret = "" Then
                ret = "<None>"
            End If
            Return ret
        End Function

        Private Sub OnListChanged()
            fRelationCollection.OnListChanged(Me)
        End Sub
    End Class
    Public Class Category
        Private fID As Integer
        Private fName As String
        Public Sub New(ByVal id As Integer, ByVal name As String)
            Me.fID = id
            Me.fName = name
        End Sub

        Public ReadOnly Property ID() As Integer
            Get
                Return fID
            End Get
        End Property

        Public Property CategoryName() As String
            Get
                Return fName
            End Get
            Set(ByVal value As String)
                fName = value
            End Set
        End Property
    End Class
    Public Class Relation
        Friend fTask As Task
        Friend fCategory As Category
        Public Sub New(ByVal task As Task, ByVal category As Category)
            Me.fTask = task
            Me.fCategory = category
        End Sub
        Public Property TaskName() As String
            Get
                Return fTask.TaskName
            End Get
            Set(ByVal value As String)
                fTask.TaskName = value
            End Set
        End Property
        Public Property DueDate() As Date
            Get
                Return fTask.DueDate
            End Get
            Set(ByVal value As Date)
                fTask.DueDate = value
            End Set
        End Property
        Public ReadOnly Property CategoryName() As String
            Get
                Return fCategory.CategoryName
            End Get
        End Property
        Public Property Complete() As Boolean
            Get
                Return fTask.Complete
            End Get
            Set(ByVal value As Boolean)
                fTask.Complete = value
            End Set
        End Property

        Public Property PercentComplete() As Integer
            Get
                Return fTask.PercentComplete
            End Get
            Set(ByVal value As Integer)
                fTask.PercentComplete = value
            End Set
        End Property

        Public Property Note() As String
            Get
                Return fTask.Note
            End Get
            Set(ByVal value As String)
                fTask.Note = value
            End Set
        End Property
    End Class
    Public Class Tasks
        Inherits BindingCollection

        Public Shared MaxTasks As Integer = 7
        Public Shared Function GetTasks(ByVal collection As TasksWithCategories) As Tasks
            Dim ret As New Tasks()
            Dim rnd As New Random()
            For i As Integer = 0 To MaxTasks - 1
                ret.List.Add(New Task(collection, i + 1, "Task" & (i + 1).ToString(), Date.Today.AddDays(rnd.Next(5))))
                If i = 2 Then
                    ret(i).PercentComplete = 50
                End If
                If i = 6 Then
                    ret(i).PercentComplete = 100
                End If
            Next i
            Return ret
        End Function
        Default Public Property Item(ByVal index As Integer) As Task
            Get
                Return DirectCast(List(index), Task)
            End Get
            Set(ByVal value As Task)
                List(index) = value
            End Set
        End Property
    End Class
    Public Class Categories
        Inherits BindingCollection

        Public Shared MaxCategories As Integer = 6
        Public Shared Function GetCategories(ByVal collection As TasksWithCategories) As Categories
            Dim ret As New Categories()
            Dim names() As String = { "Business", "Competitor", "Favorites", "Gifts", "Goals", "Holiday", "Ideas", "International", "Personal" }
            For i As Integer = 0 To names.Length - 1
                ret.List.Add(New Category(i + 1, names(i)))
            Next i
            Return ret
        End Function
        Default Public Property Item(ByVal index As Integer) As Category
            Get
                Return DirectCast(List(index), Category)
            End Get
            Set(ByVal value As Category)
                List(index) = value
            End Set
        End Property
    End Class
    Public Class TasksWithCategories
        Inherits BindingCollection

        Friend fTasks As Tasks
        Friend fCategories As Categories
        Public Sub New()
            fTasks = Tasks.GetTasks(Me)
            fCategories = Categories.GetCategories(Me)
        End Sub
        Public Shared Function GetTasksWithCategories() As TasksWithCategories
            Dim ret As New TasksWithCategories()
            Dim rnd As New Random()
            For i As Integer = 0 To Tasks.MaxTasks - 1
                For j As Integer = 0 To (1 + rnd.Next(Categories.MaxCategories)) - 1
                    Dim cat As Category = ret.fCategories(rnd.Next(ret.fCategories.Count))
                    If Not ret.HasCategory(ret.fTasks(i), cat) Then
                        ret.List.Add(New Relation(ret.fTasks(i), cat))
                    End If
                Next j
            Next i
            Return ret
        End Function
        Default Public Property Item(ByVal index As Integer) As Relation
            Get
                Return DirectCast(List(index), Relation)
            End Get
            Set(ByVal value As Relation)
                List(index) = value
            End Set
        End Property
        Public Overrides Sub OnListChanged(ByVal item As Object)
            If item Is Nothing Then
                Return
            End If
            For i As Integer = 0 To Me.Count - 1
                If item.Equals(Me(i).fTask) Then
                    Me.OnListChanged(New ListChangedEventArgs(ListChangedType.ItemChanged, i))
                End If
            Next i
        End Sub
        Public Function HasCategory(ByVal task As Task, ByVal category As Category) As Boolean
            For i As Integer = 0 To Me.Count - 1
                If Me(i).fCategory Is category AndAlso Me(i).fTask Is task Then
                    Return True
                End If
            Next i
            Return False
        End Function
    End Class
End Namespace
